### PR TITLE
Refactor VM classes

### PR DESCRIPTION
**Pull request by ehankland**
Add a BaseOsMixin and a LinuxMixin subclass that holds linux specific methods and attributes

### CODE REVIEW

**Comment by ehankland** [comment]
test this please.

**Comment by ehankland** [comment]
test this please.

File: perfkitbenchmarker/benchmark_spec.py [hunk 1]
```diff
@@ -109,6 +109,11 @@
 
 flags.DEFINE_enum('cloud', GCP, [GCP, AZURE, AWS, DIGITALOCEAN],
                   'Name of the cloud to use.')
+flags.DEFINE_enum('os_type', DEBIAN,
+                  [DEBIAN, RHEL],
+                  'The VM\'s OS type. For Linux variants, this includes OSs '
+                  'which are based on the OS (i.e. Ubuntu\'s os_type is '
```
Comment (cmccoy) [inline perfkitbenchmarker/benchmark_spec.py]: "OSs based on the OS" is a bit confusing. Could you clarify?

File: perfkitbenchmarker/benchmark_spec.py [hunk 2]
```diff
@@ -109,6 +109,11 @@
...
+                  'Debian). This will determine the OS Mixin class used.')
```
Comment (cmccoy) [inline perfkitbenchmarker/benchmark_spec.py]: Command-line users won't know what an OS Mixin is. Package manager?

File: perfkitbenchmarker/benchmark_spec.py [hunk 3]
```diff
@@ -298,7 +303,8 @@ def PrepareVm(self, vm, firewall):
     vm.Create()
     logging.info('VM: %s', vm.ip_address)
     logging.info('Waiting for boot completion.')
-    firewall.AllowPort(vm, vm.ssh_port)
+    if vm.ssh_port:
```
Comment (cmccoy) [inline perfkitbenchmarker/benchmark_spec.py]: If this is in preparation for windows support, maybe VMs should expose a set of ports required for remote access?

File: perfkitbenchmarker/virtual_machine.py [hunk 4]
```diff
@@ -108,9 +97,8 @@ def __init__(self, vm_spec):
     """
     super(BaseVirtualMachine, self).__init__()
     with self._instance_counter_lock:
-      self.name = 'perfkit-%s-%d' % (FLAGS.run_uri, self._instance_counter)
+      self.name = 'pkb-%s-%d' % (FLAGS.run_uri, self._instance_counter)
```
Comment (cmccoy) [inline perfkitbenchmarker/virtual_machine.py]: Could you make a separate PR for this? I'm used to looking for resources with a `perfkit-` prefix, others may be as well.

File: perfkitbenchmarker/virtual_machine.py [hunk 5]
```diff
@@ -152,99 +129,101 @@ def CreateScratchDisk(self, disk_spec):
...
+
+  def __init__(self):
+    super(BaseOsMixin, self).__init__()
+    self._installed_packages = set()
+
+    self.bootable_time = None
+    self.hostname = None
+
+    # Ports that may be opened by benchmark_spec
+    self.ssh_port = None
```
Comment (cmccoy) [inline perfkitbenchmarker/virtual_machine.py]: This seems Linux-specific.

File: perfkitbenchmarker/linux_virtual_machine.py [hunk 6]
```diff
@@ -0,0 +1,606 @@
...
+FLAGS = flags.FLAGS
+
+
+class BaseLinuxMixin(virtual_machine.BaseOsMixin):
+  """Class that holds Linux related VM methods and attributes."""
+
+  # If multiple ssh calls are made in parallel using -t it will mess
+  # the stty settings up and the terminal will become very hard to use.
+  # Serializing calls to ssh with the -t option fixes the problem.
+  pseudo_tty_lock = threading.Lock()
```
Comment (cmccoy) [inline perfkitbenchmarker/linux_virtual_machine.py]: Private?

File: perfkitbenchmarker/virtual_machine.py [hunk 7]
```diff
@@ -152,99 +129,101 @@ def CreateScratchDisk(self, disk_spec):
...
+
+    # Cached values
+    self._reachable = {}
+    self._total_memory_kb = None
+    self._num_cpus = None
+
+  @abc.abstractmethod
+  def RemoteCommand(self, command, should_log=False, ignore_failure=False,
+                    suppress_warning=False, **kwargs):
```
Comment (cmccoy) [inline perfkitbenchmarker/virtual_machine.py]: Document or remove `**kwargs`.

Comment (cmccoy) [inline perfkitbenchmarker/virtual_machine.py]: Sorry, you do below. nm.

File: perfkitbenchmarker/virtual_machine.py [hunk 8]
```diff
@@ -254,123 +233,48 @@ def RemoteCopy(self, file_path, remote_path='', copy_to=True):
...
 
-    Args:
-      command: A valid bash command.
+  @abc.abstractmethod
+  def WaitForBootCompletion(self):
+    """Waits until VM is has booted.
 
-    Returns:
-      A tuple of stdout and stderr from running the command.
+    Implmentations of this method should set the 'bootable_time' attribute
```
Comment (cmccoy) [inline perfkitbenchmarker/virtual_machine.py]: Spelling: Implementations.

File: perfkitbenchmarker/virtual_machine.py [hunk 9]
```diff
@@ -254,123 +233,48 @@ def RemoteCopy(self, file_path, remote_path='', copy_to=True):
...
-      should_log: A boolean indicating whether the command result should be
-          logged at the info level. Even if it is false, the results will
-          still be logged at the debug level.
-      retries: The maximum number of times RemoteCommand should retry SSHing
-          when it receives a 255 return code.
-      ignore_failure: Ignore any failure if set to true.
-      login_shell: Run command in a login shell.
-      suppress_warning: Suppress the result logging from IssueCommand when the
-          return code is non-zero.
+  def Startup(self):
```
Comment (cmccoy) [inline perfkitbenchmarker/virtual_machine.py]: Maybe `OnStartup`, to avoid confusion with starting halted VMs?

File: perfkitbenchmarker/benchmark_spec.py (hunk 1 above)
Comment (ehankland) [inline perfkitbenchmarker/benchmark_spec.py]: Reworded flag documentation to make it more understandable.

File: perfkitbenchmarker/benchmark_spec.py (hunk 2 above)
Comment (ehankland) [inline perfkitbenchmarker/benchmark_spec.py]: Removed.

File: perfkitbenchmarker/benchmark_spec.py (hunk 3 above)
Comment (ehankland) [inline perfkitbenchmarker/benchmark_spec.py]: Modified.

File: perfkitbenchmarker/virtual_machine.py (hunk 4 above)
Comment (ehankland) [inline perfkitbenchmarker/virtual_machine.py]: Sure - I'll put it off till the next PR. The issue is that on Azure they use the vm name as the username and they have a length restriction of 15 characters for the username.

File: perfkitbenchmarker/virtual_machine.py (hunk 5 above)
Comment (ehankland) [inline perfkitbenchmarker/virtual_machine.py]: Changed to be more OS agnostic

File: perfkitbenchmarker/linux_virtual_machine.py (hunk 6 above)
Comment (ehankland) [inline perfkitbenchmarker/linux_virtual_machine.py]: Done

File: perfkitbenchmarker/virtual_machine.py (hunk 8 above)
Comment (ehankland) [inline perfkitbenchmarker/virtual_machine.py]: Done

File: perfkitbenchmarker/virtual_machine.py (hunk 9 above)
Comment (ehankland) [inline perfkitbenchmarker/virtual_machine.py]: Done.

**Comment by ehankland** [comment]
test this please.

**Comment by ehankland** [comment]
PTAL

**Comment by cmccoy** [comment]
Probably not something to be addressed by this PR, but having to define (empty) classes for every combination of (package manager, cloud) seems a little verbose. Could we make the package manager an attribute of `<Cloud>LinuxVirtualMachine`?

**Comment by ehankland** [comment]
Yeah - that wouldn't be too hard. I'd like to leave it for after the rest of this refactor though.

File: perfkitbenchmarker/linux_virtual_machine.py [hunk 10]
```diff
@@ -0,0 +1,693 @@
...
+    stdout, stderr, retcode = vm_util.IssueCommand(scp_cmd, timeout=None)
+
+    if retcode:
+      full_cmd = ' '.join(scp_cmd)
+      error_text = ('Got non-zero return code (%s) executing %s\n'
+                    'STDOUT: %sSTDERR: %s' %
+                    (retcode, full_cmd, stdout, stderr))
+      raise errors.VirtualMachine.RemoteCommandError(error_text)
+
+  def LongRunningRemoteCommand(self, command):
```
Comment (cmccoy) [inline perfkitbenchmarker/linux_virtual_machine.py]: This was removed by #310.

Comment (ehankland) [inline perfkitbenchmarker/linux_virtual_machine.py]: oops - missed that - sorry

File: perfkitbenchmarker/virtual_machine.py [hunk 11]
```diff
@@ -181,99 +128,101 @@ def CreateScratchDisk(self, disk_spec):
...
-               'sudo chown -R $USER:$USER {1};').format(device_path, mount_path)
-    self.RemoteCommand(mnt_cmd)
+class BaseOsMixin(object):
+  """The base class for OS Mixin classes.
 
-  def RenderTemplate(self, template_path, remote_path, context):
-    """Renders a local Jinja2 template and copies it to the remote host.
+  This class holds VM methods and attributes relating to the VM's guest OS.
+  For methods and attributes that relate to the VM as a cloud resource,
+  see BaseVirtualMachine and its subclasses.
```
Comment (cmccoy) [inline perfkitbenchmarker/virtual_machine.py]: Could you document the public attributes of this class (`bootable_time, hostname, remote_access_ports`)?

**Comment by cmccoy** [comment]
Code LGTM. Have you tested against CentOS?

File: perfkitbenchmarker/linux_virtual_machine.py [hunk 12]
```diff
@@ -0,0 +1,662 @@
...
+
+    Raises:
+      RemoteCommandError: If there was a problem copying the file.
+    """
+    if vm_util.RunningOnWindows():
+      if ':' in file_path:
+        # scp doesn't like colons in paths.
+        file_path = file_path.split(':', 1)[1]
+      # Replace the last instance of '\' with '/' to make scp happy.
+      file_path = '/'.join(file_path.rsplit('\\', 1))
```
Comment (ivansmf) [inline perfkitbenchmarker/linux_virtual_machine.py]: You might want to consider os.path.sep instead of hardcoding the separator.

File: perfkitbenchmarker/linux_virtual_machine.py [hunk 13]
```diff
@@ -0,0 +1,662 @@
...
+      remote_path: Optional path of where to copy file on remote host.
+      copy_to: True to copy to vm, False to copy from vm.
+
+    Raises:
+      RemoteCommandError: If there was a problem copying the file.
+    """
+    if vm_util.RunningOnWindows():
+      if ':' in file_path:
+        # scp doesn't like colons in paths.
+        file_path = file_path.split(':', 1)[1]
```
Comment (ivansmf) [inline perfkitbenchmarker/linux_virtual_machine.py]: You can use drive, directory = os.path.splitdrive(clean_path)

File: perfkitbenchmarker/linux_virtual_machine.py [hunk 14]
```diff
@@ -0,0 +1,662 @@
...
+    Returns:
+      A tuple of stdout and stderr from running the command.
+
+    Raises:
+      RemoteCommandError: If there was a problem establishing the connection.
+    """
+    if vm_util.RunningOnWindows():
+      # Multi-line commands passed to ssh won't work on Windows unless the
+      # newlines are escaped.
+      command = command.replace('\n', '\\n')
```
Comment (ivansmf) [inline perfkitbenchmarker/linux_virtual_machine.py]: You can use os.linesep instead

File: perfkitbenchmarker/linux_virtual_machine.py [hunk 15]
```diff
@@ -0,0 +1,662 @@
...
+    if copy_to:
+      scp_cmd.extend([file_path, remote_location])
+    else:
+      scp_cmd.extend([remote_location, file_path])
+
+    stdout, stderr, retcode = vm_util.IssueCommand(scp_cmd, timeout=None)
+
+    if retcode:
+      full_cmd = ' '.join(scp_cmd)
+      error_text = ('Got non-zero return code (%s) executing %s\n'
```
Comment (ivansmf) [inline perfkitbenchmarker/linux_virtual_machine.py]: use os.linesep

File: perfkitbenchmarker/linux_virtual_machine.py [hunk 16]
```diff
@@ -0,0 +1,662 @@
...
+          break
+    finally:
+      if login_shell:
+        self._pseudo_tty_lock.release()
+
+    if retcode:
+      full_cmd = ' '.join(ssh_cmd)
+      error_text = ('Got non-zero return code (%s) executing %s\n'
+                    'Full command: %s\nSTDOUT: %sSTDERR: %s' %
+                    (retcode, command, full_cmd, stdout, stderr))
```
Comment (ivansmf) [inline perfkitbenchmarker/linux_virtual_machine.py]: Same for all \n, use os.linesep instead
